nestadia/src/apu/mod.rs

```diff
@@ -0,0 +1,257 @@
+use alloc::vec::{Drain, Vec};
+use bitflags::bitflags;
+
+mod common;
+mod noise;
+mod pulse;
+mod triangle;
+
+use self::common::SequenceMode;
+use self::noise::NoiseChannel;
+use self::pulse::PulseChannel;
+use self::triangle::TriangleChannel;
+
+const PULSE_MIXING_TABLE: [f32; 31] = {
+    let mut table = [0f32; 31];
+    let mut i = 1;
+    while i < 31 {
+        table[i as usize] = 95.52 / (8128.0 / i as f32 + 100.0);
+        i += 1;
+    }
+    table
+};
+
+const TND_MIXING_TABLE: [f32; 203] = {
+    let mut table = [0f32; 203];
+    let mut i = 1;
+    while i < 203 {
+        table[i as usize] = 163.67 / (24329.0 / i as f32 + 100.0);
+        i += 1;
+    }
+    table
+};
+
+const MAX_SAMPLES: usize = 1024;
+const SAMPLE_RATE: f32 = 44100.0;
+const CPU_FREQUENCY: f32 = 1789773.0;
+const CPU_CYCLES_PER_SAMPLE: u16 = (CPU_FREQUENCY / SAMPLE_RATE) as u16;
+
+bitflags! {
+    struct ChannelEnable: u8 {
+        const PULSE1_ENABLE = 0b00000001;
+        const PULSE2_ENABLE = 0b00000010;
+        const TRIANGLE_ENABLE = 0b00000100;
+        const NOISE_ENABLE = 0b00001000;
+        const DMC_ENABLE = 0b00010000;
+    }
+}
+
+pub struct Apu {
+    // Channels
+    pulse_channel_1: PulseChannel,
+    pulse_channel_2: PulseChannel,
+    triangle_channel: TriangleChannel,
+    noise_channel: NoiseChannel,
+
+    // Frame counter
+    disable_interrupts: bool,
+    sequence_mode: SequenceMode,
+    frame_counter: u16,
+    cycle_count: u16,
+
+    // Sampling
+    sample_sum: f32,
+    sample_count: u16,
+    samples: Vec<i16>,
+}
+
+impl Default for Apu {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Apu {
+    pub fn new() -> Self {
+        Self {
+            pulse_channel_1: PulseChannel::new(true),
+            pulse_channel_2: PulseChannel::new(false),
+            triangle_channel: Default::default(),
+            noise_channel: Default::default(),
+
+            disable_interrupts: false,
+            sequence_mode: Default::default(),
+            frame_counter: 0,
+            cycle_count: 0,
+
+            sample_sum: 0.0,
+            sample_count: 0,
+            samples: Vec::with_capacity(MAX_SAMPLES),
+        }
+    }
+
+    pub fn reset(&mut self) {
+        *self = Default::default();
+    }
+
+    pub fn write(&mut self, _bus: /*&mut PpuBus<'_>*/ u8, addr: u16, data: u8) {
```

Comment:
What is this with the bus?

Reply:
Ah... forgot to remove that. The APU will probably have a bus for the DMC channel, but right now that was just a placeholder. I'll remove it.